### PR TITLE
Port test_checker.py to Python 3

### PR DESCRIPTION
**Pull request by itamarst**
Fixes https://tahoe-lafs.org/trac/tahoe-lafs/ticket/3459

### CODE REVIEW

**Comment by itamarst** [comment]
I was hoping this would provide coverage of allmydata.immutable.checker but apparently not. Guess I'll keep looking :(

File: src/allmydata/immutable/filenode.py [hunk 1]
```diff
@@ -139,7 +139,7 @@ def _gather_repair_results(self, ur, cr, crr):
             for server in servers:
                 sm.add(shnum, server)
                 servers_responding.add(server)
-        servers_responding = sorted(servers_responding)
+        servers_responding = sorted(servers_responding, key=lambda o: id(o))
```
Comment (itamarst) [inline src/allmydata/immutable/filenode.py:142]: Not sure why the sort was here to begin with... objects in Python 3 aren't sortable if they're not explicitly comparable, and these aren't. Other option is just get rid of the sorting.

File: src/allmydata/web/common.py [hunk 2]
```diff
@@ -1,45 +1,32 @@
```
Comment (itamarst) [inline src/allmydata/web/common.py:1]: Changes to this module are just moving stuff into common_py3.py so nevow imports don't get in the way.

File: src/allmydata/web/operations.py [hunk 3]
```diff
@@ -142,36 +143,3 @@ def _release_ophandle(self, ophandle):
             self.timers[ophandle].cancel()
         self.timers.pop(ophandle, None)
         self.handles.pop(ophandle, None)
-
```
Comment (itamarst) [inline src/allmydata/web/operations.py:145]: Moved into common_py3.py.

File: src/allmydata/web/common.py (hunk 2 above)
Comment (chadwhitacre) [reply src/allmydata/web/common.py:1]: The `nevow` imports in `common.py` look minimal. Could we put those behind an `if PY2` in order to leave the rest of `common.py` intact? That would save us from having to fold (even more of) `common_py3.py` back in later (sez the guy who just did [that](https://github.com/tahoe-lafs/tahoe-lafs/pull/842) for `test.common_py3` 😬 ). Maybe something like:

```python3
if PY2:
    from nevow import appserver
    from nevow.inevow import IRequest

    class MyExceptionHandler(appserver.DefaultExceptionHandler, object):                                                                        
        def simple(self, ctx, text, code=http.BAD_REQUEST):
            pass # full version
        def renderHTTP_exception(self, ctx, f):
            pass # full version

    def get_root():
            pass # full version

elif PY3:

    class NotPortedToPython3(NotImplementedError):
         pass

    class MyExceptionHandler(appserver.DefaultExceptionHandler, object):                                                                        
        def simple(self, ctx, text, code=http.BAD_REQUEST):
            raise NotPortedToPython3
        def renderHTTP_exception(self, ctx, f):
            raise NotPortedToPython3

    def get_root():
            raise NotPortedToPython3
    
else:
    raise heck
```

File: src/allmydata/test/test_util.py [hunk 4]
```diff
@@ -13,16 +13,19 @@
 import six
 import os, time, sys
 import yaml
+import json
 
 from twisted.trial import unittest
 
 from allmydata.util import idlib, mathutil
 from allmydata.util import fileutil
+from allmydata.util import jsonbytes
```
Comment (chadwhitacre) [inline src/allmydata/test/test_util.py:22]: `jsonbytes` would seem a good candidate for a separate PR to aid digestion, if desired.

**Review by chadwhitacre — CHANGES_REQUESTED** [review]

**Comment by itamarst** [comment]
Good point re common.py, will switch it back and hopefully PR will then be small enough you can look at jsonbytes in same PR.

File: src/allmydata/web/common.py (hunk 2 above)
Comment (itamarst) [reply src/allmydata/web/common.py:3]: Good idea, done.

File: src/allmydata/mutable/publish.py [hunk 5]
```diff
@@ -526,7 +526,7 @@ def publish(self, newdata):
         return self.done_deferred
 
     def _get_some_writer(self):
-        return list(self.writers.values()[0])[0]
+        return list(list(self.writers.values())[0])[0]
```
Comment (chadwhitacre) [inline src/allmydata/mutable/publish.py:529]: Yikes.

**Review by chadwhitacre — APPROVED** [review]

File: src/allmydata/immutable/filenode.py (hunk 1 above)
Comment (meejah) [reply src/allmydata/immutable/filenode.py:155]: I can't tell why this is sorted either.
I'd be fine with removing it if the tests still pass (perhaps it was an attempt to make testing easier/more-predictable? if it affects output order somehow?)

That said, there are certainly cases where the order matters for servers .. but I don't think this is one (and if it were, they should be sorted on their "server key" or something, not where they landed in memory).

File: src/allmydata/mutable/publish.py (hunk 5 above)
Comment (meejah) [reply src/allmydata/mutable/publish.py:529]: yeah..."legacy code gunna legacy" i guess :woman_shrugging:

File: src/allmydata/mutable/publish.py [hunk 6]
```diff
@@ -704,7 +704,7 @@ def _encode_segment(self, segnum):
         self.log("Pushing segment %d of %d" % (segnum + 1, self.num_segments))
         data = self.data.read(segsize)
         # XXX: This is dumb. Why return a list?
-        data = "".join(data)
+        data = b"".join(data)
```
Comment (meejah) [inline src/allmydata/mutable/publish.py:707]: Is that XXX comment irrelevant? Not really part of this PR but ... it's not returning a list :)

**Review by meejah — APPROVED** [review]
LGTM.

A version of the "jsonbytes" that that confirms pass-through of unicode could improve the test-coverage.

File: src/allmydata/check_results.py [hunk 7]
```diff
@@ -1,3 +1,4 @@
+from past.builtins import unicode
```
Comment (chadwhitacre) [inline src/allmydata/check_results.py:1]: I was thinking about this last night. It seems non-ideal to maintain old names, but I guess the goal is not to remove Python 2 support, but rather to add Python 3 support.

Comment (itamarst) [reply src/allmydata/check_results.py:1]: This doesn't have to stay this way—this module isn't being ported yet, so trying to do minimal, least risky changes just to get tests elsewhere passing. When it gets ported, yeah, might be worth removing it.

File: src/allmydata/mutable/publish.py (hunk 6 above)
Comment (itamarst) [reply src/allmydata/mutable/publish.py:707]: Hm. Maybe it's commenting on `self.data.read()`?

File: src/allmydata/check_results.py (hunk 7 above)
Comment (wearpants) [reply src/allmydata/check_results.py:1]: > the goal is not to remove Python 2 support, but rather to add Python 3 support.

exactly this

**Comment by itamarst** [comment]
Looks like the pypy failure is the common intermittent one, so just going to merge.
